### ISSUE TITLE
Search using filters not working correctly

### ISSUE THREAD

**Issue by roparz** [meilisearch/meilisearch#3178]
**Describe the bug**
I have an index of users with ~250k documents.
All documents are correctly imported (number of documents in /stats reflects the count in my DB).
When I search for users using filters, with different values, sometimes it returns the expected documents, sometimes an empty result.

**To Reproduce**
1. With an index of ~250k documents looking like the following:
```
{
  "fullname":"xxxxxx",
  "email":"xxxxx",
  "company_number":"0123456789",
  "emails":["xxxxxx"],
  "id":246500,
  "firstname":"xxx",
  "lastname":"xxx",
  "fk_active_company":273, <= ~600 différent values
  "email_verified_at":null,
  "state":"guest",
  "status":"employee",
  "company":"xxx",
  "is_deleted":false
}
```
2. GET a user using it's ID. It returns the document.
```
curl -X GET "http://localhost:7700/indexes/users/documents/246500"
```
3. Search for users using `fk_active_company` filter. Result is empty.
```
curl \
  -H "Content-Type: application/json"\
  -X POST "http://localhost:7700/indexes/users/search" \
  --data-binary '{ "filter": [ "fk_active_company = 273" ] }'

{"hits":[],"query":"","processingTimeMs":0,"limit":20,"offset":0,"estimatedTotalHits":0}
```

**Expected behavior**
Search with filters should work at any time.

**Meilisearch version: v0.30.0**

**Additional context**
Infra: AWS Meilisearch AMI v0.30.0 on a `c5.large` instance.

---

# TODO

- [ ] Implement changes in [Milli](https://github.com/meilisearch/milli/)
- [ ] Release a Milli version containing these changes
- [ ] Bump this new Milli version in Meilisearch and merge it into `main`

**Comment by ManyTheFish** [comment]
Elevate this issue from `support` to `bug` because @imkarthikk faced the same problem.

@imkarthikk, don't hesitate to give us additional information about your own issue.

related to discord support discussion: https://discord.com/channels/1006923006964154428/1047923325093019768/1047923325093019768

**Comment by loiclec** [comment]
I'll try to double-check today to be sure, but I think this bug is caused by the same "corrupted database" issue causing https://github.com/meilisearch/meilisearch/issues/3165 and fixed by https://github.com/meilisearch/milli/pull/712

**Comment by curquiza** [comment]
Fixed by #3202, will be released in v0.30.1 today!

**Comment by loiclec** [comment]
Hi everyone, I am very sorry, but the update to v0.30.1 did not *completely* fix this issue. Most of the missing filter results were indeed caused by a corrupted database, as I suspected. However, I then found another bug in the search algorithms that could cause a filter to miss some documents. We will quickly prepare a v0.30.2 (sigh...) containing a fix for this new bug as well. Thank you so much for the bug reports, we will try to do better in the future ❤️
